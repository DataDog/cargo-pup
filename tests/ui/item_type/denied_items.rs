//@compile-flags: --crate-name test
//@compile-flags: --crate-type lib

// This struct should trigger the lint (denied item type)
pub struct ForbiddenStruct { //~ ERROR: struct 'ForbiddenStruct' is not allowed in this module
    pub field: String,
}

// This enum should trigger the lint (denied item type)
pub enum ForbiddenEnum { //~ ERROR: enum 'ForbiddenEnum' is not allowed in this module
    VariantA,
    VariantB,
}

// This trait should trigger the lint (denied item type)
pub trait ForbiddenTrait { //~ ERROR: trait 'ForbiddenTrait' is not allowed in this module
    fn method(&self);
}

// This nested module should trigger the lint (denied item type)
pub mod nested_module { //~ ERROR: module 'nested_module' is not allowed in this module
    pub fn allowed_function() {
        println!("This function is inside a forbidden module!");
    }
}

// These items should be allowed (not in denied_items list)
pub fn allowed_function() {
    println!("This function is allowed!");
}

pub const ALLOWED_CONST: i32 = 42;

pub static ALLOWED_STATIC: &str = "This static is allowed";

// Impl blocks not for a trait should be allowed
impl ForbiddenStruct {
    pub fn new(value: String) -> Self {
        Self { field: value }
    }
}