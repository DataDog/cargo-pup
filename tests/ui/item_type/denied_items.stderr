error: struct 'ForbiddenStruct' is not allowed in this module
  --> tests/ui/item_type/denied_items.rs:5:1
   |
LL | / pub struct ForbiddenStruct {
LL | |     pub field: String,
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'deny_item_types_test'.
   = note: `#[deny(item_type)]` on by default

error: enum 'ForbiddenEnum' is not allowed in this module
  --> tests/ui/item_type/denied_items.rs:10:1
   |
LL | / pub enum ForbiddenEnum {
LL | |     VariantA,
LL | |     VariantB,
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'deny_item_types_test'.

error: trait 'ForbiddenTrait' is not allowed in this module
  --> tests/ui/item_type/denied_items.rs:16:1
   |
LL | / pub trait ForbiddenTrait {
LL | |     fn method(&self);
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'deny_item_types_test'.

error: module 'nested_module' is not allowed in this module
  --> tests/ui/item_type/denied_items.rs:21:1
   |
LL | / pub mod nested_module {
LL | |     pub fn allowed_function() {
LL | |         println!("This function is inside a forbidden module!");
LL | |     }
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'deny_item_types_test'.

error: aborting due to 4 previous errors

